syntax = "proto3";

package ai_generation;

// Content entry input for quiz generation
message ContentEntry {
    int64 id = 1;
    string page_title = 2;
    string content = 3;
    optional int32 word_count_analyzed = 4;
}

// Request message for generating quiz questions
message GenerateQuizRequest {
    string instructions = 1;
    repeated ContentEntry content_entries = 2;
}

// Request message for generating topics
message GenerateTopicsRequest {
    string content = 1;
    string page_title = 2;
    repeated string existing_topics = 3;
}

// Response message for generated topics
message GenerateTopicsResponse {
    repeated string topics = 1;
}

// Question option definition
message QuestionOption {
    string option_text = 1;
    string option_explanation = 2;
    bool is_correct = 3;
}

// Question definition
message Question {
    string question = 1;
    string type = 2;
    repeated QuestionOption options = 3;
}

// Progress message for each content entry
message QuizGenerationProgress {
    oneof message_type {
        GenerationStatus status = 1;
        bool completed = 2;
        GenerationResult result = 3;
    }
}

// Status message indicating which content entry is being processed
message GenerationStatus {
    int64 content_entry_id = 1;
    string page_title = 2;
    string status = 3; // e.g., "Generating questions for page title X"
    optional int32 word_count_analyzed = 4;
}

// Result message containing generated questions for a content entry
message GenerationResult {
    int64 content_entry_id = 1;
    string page_title = 2;
    repeated Question questions = 4;
}

// Quiz Generation Service
service AiGenerationService {
    rpc GenerateQuiz(GenerateQuizRequest) returns (stream QuizGenerationProgress);
    rpc GenerateTopics(GenerateTopicsRequest) returns (GenerateTopicsResponse);
}